pkg/pubsub/adapter/adapter.go

```diff
@@ -18,253 +18,252 @@ package adapter
-	// Apply CloudEvent override extensions to the outbound event.
-	for k, v := range a.extensions {
-		event.SetExtension(k, v)
+	// Only if the message is not from a reply, then we should add the override extensions.
+	if !reply {
+		// Apply CloudEvent override extensions to the outbound event.
+		// This code will be mainly executed by Sources.
+		for k, v := range a.args.Extensions {
+			event.SetExtension(k, v)
+		}
 	}
 
-	// Send the event and report the count.
-	rctx, r, err := a.outbound.Send(ctx, event)
-	rtctx := cloudevents.HTTPTransportContextFrom(rctx)
-	a.reporter.ReportEventCount(args, rtctx.StatusCode)
+	response, err := a.sendMsg(ctx, a.args.SinkURI, (*binding.EventMessage)(event))
 	if err != nil {
-		return err
-	} else if r != nil {
-		resp.RespondWith(nethttp.StatusOK, r)
+		a.logger.Error("Failed to send message to sink", zap.String("address", a.args.SinkURI), zap.Error(err))
+		msg.Nack()
+		return
 	}
-	return nil
-}
 
-func (a *Adapter) convert(ctx context.Context, m transport.Message, err error) (*cloudevents.Event, error) {
-	logger := logging.FromContext(ctx)
-	logger.Debug("Converting event from transport.")
+	defer func() {
+		if err := response.Body.Close(); err != nil {
+			a.logger.Warn("Failed to close response body", zap.Error(err))
+		}
+	}()
 
-	if msg, ok := m.(*cepubsub.Message); ok {
-		return converters.Convert(ctx, msg, a.SendMode, a.AdapterType)
-	}
-	return nil, err
-}
+	a.reporter.ReportEventCount(args, response.StatusCode)
 
-func (a *Adapter) newPubSubClient(ctx context.Context) (cloudevents.Client, error) {
-	tOpts := []cepubsub.Option{
-		cepubsub.WithProjectID(a.Project),
-		cepubsub.WithTopicID(a.Topic),
-		cepubsub.WithSubscriptionAndTopicID(a.Subscription, a.Topic),
+	if response.StatusCode/100 != 2 {
+		a.logger.Error("Event delivery failed", zap.Int("StatusCode", response.StatusCode))
+		msg.Nack()
+		return
 	}
 
-	// Make a pubsub transport for the CloudEvents client.
-	t, err := cepubsub.New(ctx, tOpts...)
+	msg.Ack()
+}
+
+func (a *Adapter) sendMsg(ctx context.Context, address string, msg binding.Message) (*nethttp.Response, error) {
+	req, err := nethttp.NewRequestWithContext(ctx, nethttp.MethodPost, address, nil)
 	if err != nil {
 		return nil, err
 	}
-
-	// Use the transport to make a new CloudEvents client.
-	return cloudevents.NewClient(t,
-		cloudevents.WithConverterFn(a.convert),
-	)
+	if err := cehttp.WriteRequest(ctx, msg, req); err != nil {
+		return nil, err
+	}
+	return a.outbound.Do(req)
 }
 
-func (a *Adapter) newHTTPClient(ctx context.Context, target string) (cloudevents.Client, error) {
-	tOpts := []http.Option{
-		cloudevents.WithTarget(target),
+func (a *Adapter) startSpan(ctx context.Context, event *cev2.Event) (context.Context, *trace.Span) {
+	spanName := tracing.SourceDestination(a.namespacedName)
+	// This receive adapter code is used both for Sources and Channels.
+	// An ugly way to identify whether it was created from a Channel is to look at the resourceGroup.
```

Comment:
yeah, this is old stuff but mainly due to metrics. At the time we wanted to know the resourceGroup of a source (e.g., cloudpubsubsources, cloudauditlogssources, etc)... so that we can report metrics from particular sources types and not just a generic `Source`, that is why this ugly resourceGroup stuff...
If you don't mind, I'd rather leave it as is for now, and then once we do a pass on metrics, we can revisit.. Sounds good?